### PR TITLE
MM-31636 Add status field

### PR DESCRIPTION
**Pull request by crspeller**
#### Summary
Adds a status field to incidents.
Status is updated at the same time that the status updates feature is used. 
The statuses are:
- Reported
- Active
- Resolved
- Archived

A follow up ticket will make Archived status actually archive the channel. 

Note: Still working on cypress tests. 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-31636

### CODE REVIEW

File: server/incident/service.go [hunk 1]
```diff
@@ -337,7 +216,7 @@ func (s *ServiceImpl) broadcastStatusUpdate(statusUpdate string, theIncident *In
 
 	duration := timeutils.DurationString(timeutils.GetTimeForMillis(theIncident.CreateAt), time.Now())
 	status := "Ongoing"
-	if !theIncident.IsActive {
+	if !theIncident.IsActive() {
 		status = "Ended"
```
Comment (cpoile) [inline server/incident/service.go:220]: This seems like it should be changed, too? (for line 224 below, in the new `service.go`)

File: server/sqlstore/migrations.go [hunk 2]
```diff
@@ -335,4 +335,26 @@ var migrations = []Migration{
 			return nil
 		},
 	},
+	{
+		fromVersion: semver.MustParse("0.5.0"),
+		toVersion:   semver.MustParse("0.6.0"),
+		migrationFunc: func(e sqlx.Ext, sqlStore *SQLStore) error {
+			if e.DriverName() == model.DATABASE_DRIVER_MYSQL {
+				if err := addColumnToMySQLTable(e, "IR_Incident", "CurrentStatus", "TEXT NOT NULL"); err != nil {
```
Comment (cpoile) [inline server/sqlstore/migrations.go:343]: I think you have MySql and Postgres reversed here

**Review by cpoile — CHANGES_REQUESTED** [review]
Nice! A few minor questions, one bigger concern (the tests indicate `LastResolvedAt` might not be working the way I expected it to work)

File: webapp/src/components/rhs/rhs_title.tsx [hunk 3]
```diff
@@ -50,7 +50,7 @@ const RHSTitle: FC = () => {
                 </Button><RHSTitleText data-testid='rhs-title'>{incident?.name || 'Incidents'}</RHSTitleText>
                 {incident && (
                     <StatusBadge
-                        isActive={incident?.is_active}
+                        isActive={incident && incidentIsActive((incident))}
```
Comment (agarciamontoro) [inline webapp/src/components/rhs/rhs_title.tsx:53]: Are these double parentheses a typo or some kind of trick I don't know?

File: server/sqlstore/incident_test.go [hunk 4]
```diff
@@ -275,13 +267,13 @@ func TestGetIncidents(t *testing.T) {
 			},
 			ExpectedErr: nil,
 		},
-		{
+		/*{
```
Comment (agarciamontoro) [inline server/sqlstore/incident_test.go:270]: Should we delete this test? And the other three that are commented out?

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
Cool! :tada: Below are mostly nits, but there are also a couple of things that got me a bit confused (I think in the same sense than Chris' comment).

Also, a couple of general questions:
- Are the CreateAt, EndAt and DeleteAt columns from the IR_Incident deprecated? Should we document this somewhere else than in the code handling it? In a comment in the migrations or somewhere else?
- Should we automatically select the `Active` status in the interactive dialog when clicking on the "Reopen Incident" button?

File: server/incident/service.go [hunk 5]
```diff
@@ -373,23 +248,32 @@ func (s *ServiceImpl) UpdateStatus(incidentID, userID string, options StatusUpda
...
 	if err = s.store.UpdateIncident(incidentToModify); err != nil {
 		return errors.Wrap(err, "failed to update incident")
 	}
 
+	if err = s.store.UpdateStatus(&SQLStatusPost{
+		IncidentID: incidentToModify.ID,
+		PostID:     post.Id,
+		Status:     options.Status,
+	}); err != nil {
+		return errors.Wrap(err, "failed to write status post to store. There is now insconsistant state.")
```
[the remark this answers is not in the document]
Comment (cpoile) [reply server/incident/service.go:270]: Are you sure we want to remove the transaction and then have this chance of inconsistent state?

File: server/api/incidents.go [hunk 6]
```diff
@@ -936,25 +910,14 @@ func parseIncidentsFilterOptions(u *url.URL) (*incident.HeaderFilterOptions, err
...
-	switch statusParam {
-	case "all", "": // default
-		status = incident.All
-	case "active":
-		status = incident.Ongoing
-	case "ended":
-		status = incident.Ended
-	default:
-		return nil, errors.Errorf("bad status parameter '%s'", statusParam)
-	}
```
[the remark this answers is not in the document]
Comment (crspeller) [reply server/api/incidents.go:950]: Removing the validation here because it's not necessary to validate the status string. Invalid just gives no results. 
But I do remember I need to change the client side of this as well. Will do.

File: server/incident/incident.go [hunk 7]
```diff
@@ -72,20 +66,62 @@ func (i *Incident) MarshalJSON() ([]byte, error) {
...
+	}
+
+	if resolvedPost == nil {
+		return 0
+	}
+
+	return resolvedPost.CreateAt
+}
+
+const (
```
[the remark this answers is not in the document]
Comment (crspeller) [reply server/incident/incident.go:113]: 0/5 I just put it with the other status stuff.

File: server/incident/incident.go [hunk 8]
```diff
@@ -72,20 +66,62 @@ func (i *Incident) MarshalJSON() ([]byte, error) {
...
+	StatusReported = "Reported"
+	StatusActive   = "Active"
+	StatusResolved = "Resolved"
+	StatusArchived = "Archived"
+)
+
 // StatusPost is information added to the incident when selecting from the db and sent to the
 // client; it is not saved to the db.
```
[the remark this answers is not in the document]
Comment (crspeller) [reply server/incident/incident.go:121]: Nope. Wasn't true before this PR ether.

File: server/incident/incident_test.go [hunk 9]
```diff
@@ -15,3 +15,202 @@ func TestIncident_MarshalJSON(t *testing.T) {
...
+					},
+					{
+						DeleteAt: 1,
+						CreateAt: 2,
+						Status:   StatusResolved,
+					},
+				},
+			},
+			expected: StatusActive,
+		},
```
[the remark this answers is not in the document]
Comment (crspeller) [reply server/incident/incident_test.go:75]: Yes it does. If it didn't, it would be inconsistent with the updates presented.

File: server/incident/incident_test.go [hunk 10]
```diff
@@ -15,3 +15,202 @@ func TestIncident_MarshalJSON(t *testing.T) {
...
+						Status:   StatusResolved,
+					},
+					{
+						DeleteAt: 0,
+						CreateAt: 456,
+						Status:   StatusResolved,
+					},
+				},
+			},
+			expected: 123,
```
[the remark this answers is not in the document]
Comment (crspeller) [reply server/incident/incident_test.go:160]: Nope 123 is correct. If you make multiple status updates with "Resolved" the actual resolution time is the first one, not the last one. 

I changed the test to have CreateAt for the first post to be 1. But the function does assume the DB has ordered these correctly, so that discrepancy is not caught.

File: server/incident/incident_test.go [hunk 11]
```diff
@@ -15,3 +15,202 @@ func TestIncident_MarshalJSON(t *testing.T) {
...
+						Status:   StatusActive,
+					},
+					{
+						DeleteAt: 0,
+						CreateAt: 456,
+						Status:   StatusResolved,
+					},
+				},
+			},
+			expected: 456,
```
[the remark this answers is not in the document]
Comment (crspeller) [reply server/incident/incident_test.go:187]: Same as above.

File: server/incident/service.go (hunk 1 above)
Comment (crspeller) [reply server/incident/service.go:220]: Yup. Good catch.

File: server/sqlstore/migrations.go (hunk 2 above)
Comment (crspeller) [reply server/sqlstore/migrations.go:343]: Good catch!

File: server/incident/incident.go [hunk 12]
```diff
@@ -72,20 +66,62 @@ func (i *Incident) MarshalJSON() ([]byte, error) {
...
+	var resolvedPost *StatusPost
+	for j := len(i.StatusPosts) - 1; j > 0; j-- {
+		if i.StatusPosts[j].DeleteAt != 0 {
+			continue
+		}
+		if i.StatusPosts[j].Status != StatusResolved && i.StatusPosts[j].Status != StatusArchived {
+			break
+		}
+		resolvedPost = &i.StatusPosts[j]
+	}
```
[the remark this answers is not in the document]
Comment (crspeller) [reply server/incident/incident.go:104]: Not the last, but the first. I think the function name might be confusing. Going to change it to just `ResolvedAt()`

File: server/command/command.go [hunk 13]
```diff
@@ -535,35 +535,7 @@ func (r *Runner) announceChannel(targetChannelID, commanderUsername, incidentCha
...
-
-	switch {
-	case errors.Is(err, incident.ErrIncidentNotActive):
-		r.postCommandResponse("This incident has already been closed.")
-		return
-	case err != nil:
-		r.warnUserAndLogErrorf("Error: %v", err)
-		return
-	}
+	r.actionUpdate()
```
[the remark this answers is not in the document]
Comment (crspeller) [reply server/command/command.go:538]: That's a great idea! https://mattermost.atlassian.net/browse/MM-32100

File: server/api/incidents.go [hunk 14]
```diff
@@ -258,10 +256,6 @@ func (h *IncidentHandler) createIncident(newIncident incident.Incident, userID s
 		return nil, errors.Wrap(incident.ErrMalformedIncident, "incident channel already has created at date")
 	}
 
-	if newIncident.EndAt != 0 {
-		return nil, errors.Wrap(incident.ErrMalformedIncident, "incident channel already has ended at date")
-	}
-
```
[the remark this answers is not in the document]
Comment (crspeller) [reply server/api/incidents.go:264]: Well it is redundant since we won't persist a given EndAt anyway. But the truth is I accidentally started removing EndAt before realizing what a core that would be.  I think we might want to clean it up and actually remove it later. Ok to keep this gone since it's redundant?

File: server/incident/service.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (crspeller) [reply server/incident/service.go:270]: What transaction? The only ones I removed had only one DB call in them?

File: webapp/src/components/rhs/rhs_footer.tsx [hunk 15]
```diff
@@ -1,25 +1,35 @@
...
 import {useDispatch} from 'react-redux';
 
 import {Footer, StyledFooterButton} from 'src/components/rhs/rhs_shared';
 import {updateStatus} from 'src/actions';
+import {Incident, incidentCurrentStatus} from 'src/types/incident';
 
-const RHSFooterSummary = () => {
+interface Props {
+    incident: Incident;
+}
```
[the remark this answers is not in the document]
Comment (crspeller) [reply webapp/src/components/rhs/rhs_footer.tsx:13]: The less components we have hooked to redux, the better. This pattern is used by other components we have.

File: webapp/src/components/rhs/rhs_title.tsx (hunk 3 above)
Comment (crspeller) [reply webapp/src/components/rhs/rhs_title.tsx:53]: It's an amazing trick that gets you to comment on my PRs 😆

File: server/sqlstore/incident_test.go (hunk 4 above)
Comment (crspeller) [reply server/sqlstore/incident_test.go:270]: Done.

File: server/incident/incident_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (cpoile) [reply server/incident/incident_test.go:75]: What happens on the front end when the user deletes a status post? Is the incident's status updated?

[the remark this answers is not in the document]
Comment (crspeller) [reply server/incident/incident_test.go:75]: Yes it is.

File: server/api/incidents.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (agarciamontoro) [reply server/api/incidents.go:264]: Oh yeah that's true, thanks for the clarification!

File: webapp/src/components/rhs/rhs_footer.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (agarciamontoro) [reply webapp/src/components/rhs/rhs_footer.tsx:13]: Ok to leave this as it is now, no problem. Simply to continue the discussion on this:

> The less components we have hooked to redux, the better

Why? I understand why we would want to keep our writes to the store to a minimum, but I don't see a problem in reading it from several places, as it is our source of truth. I wanna hear your thoughts on this!

File: webapp/src/components/rhs/rhs_title.tsx (hunk 3 above)
Comment (agarciamontoro) [reply webapp/src/components/rhs/rhs_title.tsx]: Interesting, I guess that the parentheses are simply treated as a grouping operator. I checked and you could have written `incidentIsActive(((((incident)))))`, it would still work. Maybe we can use that to emphasize some pieces of code that we find very very important :joy:

File: server/incident/incident.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (agarciamontoro) [reply server/incident/incident.go:111]: Oh you're going backwards, that's right. What if the incident was reopened?

File: server/sqlstore/incident.go [hunk 16]
```diff
@@ -205,50 +198,37 @@ func (s *incidentStore) CreateIncident(newIncident *incident.Incident) (out *inc
...
+			"TeamID":                  rawIncident.TeamID,
+			"ChannelID":               rawIncident.ChannelID,
 			"PostID":                  rawIncident.PostID,
 			"PlaybookID":              rawIncident.PlaybookID,
 			"ChecklistsJSON":          rawIncident.ChecklistsJSON,
 			"ReminderPostID":          rawIncident.ReminderPostID,
 			"PreviousReminder":        rawIncident.PreviousReminder,
 			"BroadcastChannelID":      rawIncident.BroadcastChannelID,
 			"ReminderMessageTemplate": rawIncident.ReminderMessageTemplate,
+			"CurrentStatus":           rawIncident.CurrentStatus(), // Added to make querying easier
```
Comment (agarciamontoro) [inline server/sqlstore/incident.go:218]: How would it be more difficult without this?

File: tests-e2e/cypress/integration/frontstage/slash_command/test_spec.js [hunk 17]
```diff
@@ -228,37 +228,6 @@ describe('slash command > test', () => {
                     // * Verify the ephemeral message warns about the parameter.
                     cy.verifyEphemeralMessage('Timestamp \'2020-1-1\' could not be parsed as a date. If you want the incident to start on January 2, 2006, the timestamp should be \'2006-01-02\'.');
                 });
-
-                it('creates an incident with the correct date in the past', () => {
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/frontstage/slash_command/test_spec.js:232]: Why is this one deleted?

File: tests-e2e/cypress/integration/frontstage/incident_rhs_list_spec.js [hunk 18]
```diff
@@ -74,7 +74,7 @@ describe('rhs incident list', () => {
     });
 
     describe('should show welcome screen', () => {
-        it('when user has no active incidents', () => {
+        /*it('when user has no active incidents', () => {
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/frontstage/incident_rhs_list_spec.js:77]: There are multiple commented tests in this file. Should we uncomment them?

File: tests-e2e/cypress/integration/frontstage/incident_creation_spec.js [hunk 19]
```diff
@@ -74,45 +74,6 @@ describe('incidents can be started', () => {
             cy.startIncidentWithSlashCommand(playbookName, incidentName);
             cy.verifyIncidentActive(teamId, incidentName);
         });
-
-        it('while viewing a group message channel', () => {
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/frontstage/incident_creation_spec.js:78]: Why are all these tests in this file deleted?

File: tests-e2e/cypress/integration/backstage/incident_list_spec.js [hunk 20]
```diff
@@ -189,16 +189,5 @@ describe('backstage incident list', () => {
             // * Verify "Previous" no longer shown
             cy.findByText('Previous').should('not.exist');
         });
-
-        it('by status', () => {
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/backstage/incident_list_spec.js:193]: We can no longer filter by status?

File: tests-e2e/cypress/integration/frontstage/incident_rhs_checklist_spec.js [hunk 21]
```diff
@@ -1,131 +0,0 @@
-// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Comment (agarciamontoro) [inline tests-e2e/cypress/integration/frontstage/incident_rhs_checklist_spec.js:1]: Why are we getting rid of these tests?

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
Thanks for the effort with the E2E tests, I see that you had some fun with them :joy:

I left some comments about tests being removed/commented out, and some other nits. Other than that, the only remaining "concern" I have is the one pointed out by @cpoile: from a user perspective, I'd find confusing that deleting a message changes the status.

File: tests-e2e/cypress/integration/backstage/incident_list_spec.js (hunk 20 above)
Comment (agarciamontoro) [reply tests-e2e/cypress/integration/backstage/incident_list_spec.js:193]: I saw that we can, but we only have the Ongoing and Ended status. Is this planned for a future PR?

**Comment by cpoile** [comment]
Re: the e2e tests, do you want to wait a bit while we try to change the docker image we're using to run the tests? That might fix some of these.

@crspeller how hard would it be to put back the tests for now? I tried doing it last night by reverting the commits, but I think I broke something by accident. If you put the tests back, I can test to see if changing the docker image will fix the weird failures. I would rather do that than remove e2e tests that /should/ be working.

File: server/incident/incident.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (crspeller) [reply server/incident/incident.go:111]: It will consider the total time from when the incident was first opened. I feel like that's the right time to consider because you should only be reopening in the case of a mistake, or the incident was not actually solved.

File: webapp/src/components/rhs/rhs_footer.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (crspeller) [reply webapp/src/components/rhs/rhs_footer.tsx:13]: The more connections you have to redux:
- The less portable your components are to share with other projects.
- Gives you less flexibility in re-use. What if I want to show an incident that's not the current one?
- Performance penalty for every connection. Redux has to notify every connected component of the change and that component has to do its did anything change check. 
- The less testable your components are. (not that we really test our components, but given our troubles with e2e testing, maybe more component testing and less e2e would be nice)

File: server/api/api.yaml [hunk 22]
```diff
@@ -78,11 +78,12 @@ paths:
           example: active
```
[the remark this answers is not in the document]
Comment (crspeller) [reply server/api/api.yaml:78]: It's not case sensitive.

File: server/sqlstore/incident.go (hunk 16 above)
Comment (crspeller) [reply server/sqlstore/incident.go:218]: Every time we read the incident we would have to do a query on the status posts table to fill this in. Not totally sure which one would be better.

**Comment by crspeller** [comment]
@agarciamontoro @cpoile To answer your questions on why I killed a bunch of tests, it's because I couldn't for the life of me get them working. (see my sanity level slowly dropping in my commit messages)

I have already spent way too much time on them and just want to get this PR in so I can stop blocking people.

**Comment by cpoile** [comment]
#434 is in!

**Review by cpoile — APPROVED** [review]
Hooray! :tada:

**Review by agarciamontoro — APPROVED** [review]
Great job, Christophers, thank you both! :)

**Comment by prapti** [comment]
In test queue.

**Comment by prapti** [comment]
So nice to see the status field pop up in the status update form :)

@crspeller @lieut-data I have one concern though that might be worth considering before merging, and please let me know if this has already been addressed before with any decision: with these changes, the incidents that were already ended with prior plugin versions show up as "Ongoing", however they can be filtered as "Resolved" in the backstage; and the ones resolved after this PR show as "Ended". This will cause confusion for the incidents that were supposed to be ended.

**Comment by crspeller** [comment]
@prapti That's not great. Let me take a look. I think I might have screwed up the migration in my flailing over the e2e tests.

**Comment by crspeller** [comment]
@prapti Latest commit should fix the upgrade issue.

**Review by prapti — APPROVED** [review]
Tested in local and cloud test servers. The issue of ended incidents showing up as 'Ongoing' is fixed with the latest commit.
Approving. Will add more edge cases in the test plan for later upgrades.
